### PR TITLE
Index-based random access for CRAM reader

### PR DESCRIPTION
**Pull request by athos**
This PR introduces index-based random access for the CRAM reader utilizing CRAI indexes.

A CRAI file, which is a gzipped TSV, contains information on each slice's location within the CRAM file (See [the CRAM specification](https://github.com/samtools/hts-specs/blob/26347448cadff3cf40982d60fe2a97f20d2543ea/CRAMv3.tex#L1851-L1929) for more details about the CRAI format itself). With this data, the CRAM reader can now jump directly to specific slices.

The test resource `medium.cram` has also been updated to make it better suited for testing random access. The original version of `medium.cram` was less suitable for this purpose because it had records from most reference sequences in a single slice. This resulted in most of the index entries pointing to the same slice, which did not effectively test the random access capabilities:

<details><summary>The content of the CRAI index file generated from the original medium.cram</summary>

```
0	546609	248644966	324	563	378717
1	67302	243032198	324	563	378717
2	1037809	196879769	324	563	378717
3	640302	190372931	324	563	378717
4	73079	180831597	324	563	378717
5	308487	170456298	324	563	378717
6	68788	158820431	324	563	378717
7	255347	145894170	324	563	378717
8	10884	141139885	324	563	378717
9	435983	135059009	324	563	378717
10	220061	134321225	324	563	378717
11	93226	133498223	324	563	378717
12	20440642	94620351	324	563	378717
13	19360454	87770561	324	563	378717
14	20506746	81793458	324	563	378717
15	98546	90073945	324	563	378717
16	291091	80789819	324	563	378717
17	526137	77426377	324	563	378717
18	312711	58608000	324	563	378717
19	279224	62315388	324	563	378717
20	9540822	38578696	324	563	378717
21	16096239	35007515	324	563	378717
22	187300	154950656	324	563	378717
23	136130	59221619	324	563	378717
-1	0	1	324	563	378717
-1	0	0	379629	171	98276
```
</details>

To address this, I've regenerated `medium.cram` by a small script using htsjdk (shared below, for future reference) to ensure it has suitably isolated slices:

```clojure
(require '[clojure.java.io :as io])
(import '[htsjdk.samtools CRAMFileWriter SamReaderFactory SAMRecord]
        '[htsjdk.samtools.cram.ref ReferenceSource]
        '[htsjdk.samtools.cram.structure CRAMEncodingStrategy]))

(let [rdr-factory (SamReaderFactory/makeDefault)
      strategy (doto (CRAMEncodingStrategy.)
                 (.setMinimumSingleReferenceSliceSize 300)
                 (.setReadsPerSlice 500)
                 (.setSlicesPerContainer 10))]
  (with-open [r (.open rdr-factory (io/file "test-resources/bam/medium.bam"))
              os (io/output-stream "test-resources/cram/medium.cram")
              w (CRAMFileWriter. strategy
                                 os
                                 nil
                                 false
                                 (ReferenceSource. (io/file "hg19.fa"))
                                 (.getFileHeader r)
                                 "test-resources/cram/medium.cram")]
    (doseq [^SAMRecord aln (iterator-seq (.iterator r))]
      (.addAlignment w aln)))))
```

<details><summary>The content of the CRAI index file generated from the new version of medium.cram</summary>

```
0	546609	205262429	324	563	22007
0	206547069	42644506	324	22570	7349
1	67302	231638920	30272	563	21618
1	232059310	11040190	30272	22181	4811
2	1037809	196879769	57294	563	17305
3	640302	190372931	75189	563	20699
4	73079	180831597	96478	563	16193
5	308487	170456298	113261	563	15229
6	68788	158820431	129080	563	18226
7	255347	145894170	147896	523	14473
8	10884	141139885	162919	520	15692
9	435983	73089895	179157	563	22898
9	73553224	61941768	179157	23461	9344
10	220061	134321225	211992	563	14596
11	93226	133498223	227178	563	14915
12	20440642	94620351	242683	563	21911
13	19360454	87770561	242683	563	21911
14	20506746	52593474	242683	563	21911
14	73755434	28544770	265183	523	22409
15	98546	54443777	265183	523	22409
15	54596598	35575893	288141	523	21355
16	291091	79862138	288141	523	21355
16	80191676	889234	310045	523	21151
17	526137	77426377	310045	523	21151
18	312711	53695504	310045	523	21151
18	54206778	4713933	331745	563	22323
19	279224	62315388	331745	563	22323
20	9540822	38578696	331745	563	22323
21	16096239	29902974	331745	563	22323
21	46285774	4817980	354657	563	23119
22	187300	154950656	354657	563	23119
23	136130	59221619	354657	563	23119
-1	0	1	354657	563	23119
-1	0	0	378365	171	23494
-1	0	0	378365	23665	23213
-1	0	0	378365	46878	23051
-1	0	0	378365	69929	23563
-1	0	0	378365	93492	24231
-1	0	0	378365	117723	24078
-1	0	0	378365	141801	23871
-1	0	0	378365	165672	24365
-1	0	0	378365	190037	12326
```
</details>

Additionally, I added some missing tests in `cljam.io.util-test` to cover changes introduced in a previous PR (6f1d1b8c78bc64f5953cd44ba6ebfe89a1349623).

### CODE REVIEW

**Review by alumi — COMMENTED** [review]
Thank you for the PR! It seems like the implementation is working well 👍 
I commented on the predicate for overlap. Please take a look.

File: src/cljam/io/cram/reader.clj [hunk 1]
```diff
@@ -119,3 +134,22 @@
...
+    (letfn [(read-fn [entries]
+              (fn [container-header bb]
+                (read-container-records rdr bb container-header entries)))
+            (step [[[^long offset entries] & more]]
+              (when offset
+                (.position ch offset)
+                (concat (read-container-with rdr (read-fn entries))
+                        (lazy-seq (step more)))))]
+      (filter #(and (= (:rname %) chr)
+                    (<= start (:pos %) end))
```
[the remark this answers is not in the document]
Comment (athos) [reply src/cljam/io/cram/reader.clj:154]: Thank you for pointing that out! Sorry, the end position completely slipped my mind 😱 I'll fix it soon.

The `end` field in the alignment records is currently not public (and `dissoc`ed), so I think we need to consider how to handle this field for the change. I will consider the tradeoffs around it a bit more.

[the remark this answers is not in the document]
Comment (athos) [reply src/cljam/io/cram/reader.clj]: In `311df61`, I have updated the overlap condition to include the `end` field.

Regarding the `end` field, I propose not `dissoc`ing it and instead semantically aligning it with the end field in SAM/BAM records. In this way, we can keep it as a public field in the CRAM records. What do you think about this?

[the remark this answers is not in the document]
Comment (alumi) [reply src/cljam/io/cram/reader.clj]: Thank you for the update! I agree with the approach to publish a new field `:end` 👍 
I confirmed that its semantics for placed unmapped reads align with the [existing BAM reader implementation](https://github.com/chrovis/cljam/blob/4699b7ff3c9f67812d13a743f641beced2094395/src/cljam/io/bam/decoder.clj#L183) and the behavior of samtools.

**Review by alumi — APPROVED** [review]

File: src/cljam/io/crai.clj [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+                  (let [[^long seq-id ^long start ^long span container-offset slice-offset size]
+                        (map #(Long/parseLong %) (str/split line #"\t"))
+                        unmapped? (neg? seq-id)]
+                    {:chr (if unmapped? "*" (:name (nth refs seq-id)))
+                     :start (if unmapped? 0 start)
+                     :end (if unmapped? 0 (+ start span))
+                     :container-offset container-offset
+                     :slice-offset slice-offset
+                     :size size})))
+           (#(intervals/index-intervals %))))))
```
[the remark this answers is not in the document]
Comment (athos) [reply src/cljam/io/crai.clj]: Oh, I missed this comment! Fixed it in 4795708.

**Review by niyarin — APPROVED** [review]
LGTM!

**Comment by athos** [comment]
Thank you all for reviewing!
I've squashed the commits into two. I'm merging this.

**Pull request merged**